### PR TITLE
Remove ListTasksResponse#setDiscoveryNodes()

### LINKED ISSUE #19772: discoveryNodes must be set before calling toXContent error when calling prepareListTasks().get()
When writing an integration test which extends `ESIntegTestCase` and trying to list existing tasks with `client().admin().cluster().prepareListTasks().get()`, I'm getting:

```
[2016-08-02 23:41:37,884][WARN ][org.elasticsearch.test.transport] [transport_client_node_sd1] Failed to receive message for action [cluster:monitor/tasks/lists]
java.lang.IllegalStateException: discoveryNodes must be set before calling toXContent with group_by=nodes
	at org.elasticsearch.action.admin.cluster.node.tasks.list.ListTasksResponse.toXContent(ListTasksResponse.java:159)
	at org.elasticsearch.common.Strings.toString(Strings.java:901)
	at org.elasticsearch.common.Strings.toString(Strings.java:887)
	at org.elasticsearch.action.admin.cluster.node.tasks.list.ListTasksResponse.toString(ListTasksResponse.java:208)
	at java.lang.String.valueOf(String.java:2994)
	at java.lang.StringBuilder.append(StringBuilder.java:131)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertVersionSerializable(ElasticsearchAssertions.java:678)
	at org.elasticsearch.test.transport.AssertingLocalTransport.handleParsedResponse(AssertingLocalTransport.java:85)
	at org.elasticsearch.transport.local.LocalTransport.handleResponse(LocalTransport.java:386)
	at org.elasticsearch.transport.local.LocalTransport.processReceivedMessage(LocalTransport.java:292)
	at org.elasticsearch.transport.local.LocalTransport.lambda$receiveMessage$0(LocalTransport.java:252)
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:450)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

### PR DESCRIPTION
**Pull request by dadoonet**
We just overwrite `toString()` method so it calls toXContent
with `group_by` = "whatever" so we don't try to group by nodes
which does not make sense in a toString() method.

We keep the old behavior for `toXContent()` method which
means that there is no impact in the REST layer but
only in logs and tests (where we call `toString()`).

Closes #19772.

### CODE REVIEW

**Comment by s1monw** [comment]
@dadoonet I think this change doesn't fix the issue but if prevents it from happening? I don't think there should be a state where we can't call toXContent in ListTasksResponse. So I think we should not merge this change. Rather open an issue and we try to fix the ListTasksResponse class to not have this invariant at all?

**Comment by nik9000** [comment]
I was thinking the right way to fix this is to make the toXContent version
that requires the extra state a separate method that takes the state. If we
want it we can use it in the rest handler but for things like toString we
use the other version. No more member variable to set.

I'm sorry for screwing the class up like that in the first place.

**Comment by dadoonet** [comment]
Thanks @nik9000! Just saw your comment after I pushed new changes.
Indeed. I'm going to fix it the way you said. Much better than what I did so far.

**Comment by nik9000** [comment]
> I'm going to fix it the way you said. Much better than what I did so far.

And it only took me two months to come up with!

**Comment by dadoonet** [comment]
@nik9000 I came with something else. I hope this what you were expecting.
REST tests are working fine locally.

**Comment by nik9000** [comment]
I really think `setDiscoveryNodes` was a mistake on my part. I was kind of hoping you'd remove it with this PR and undo my mistake.

**Comment by dadoonet** [comment]
@nik9000 pushed a new commit. Let me know what you think.

**Comment by dadoonet** [comment]
@nik9000 I updated the PR and update with latest changes in master.
Note that I had to add `builder.startObject();` and `builder.endObject();` because this is failing now in my tests.
Note that it was not failing before I merged master branch into this branch.

**Comment by dadoonet** [comment]
After running tests, adding `builder.startObject();` was actually a bad idea... Need to fix that and send another commit.

**Comment by dadoonet** [comment]
Well apparently, it does not come from my PR. I added a simple call to toString() in master branch: 

```
Index: core/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/TransportTasksActionTests.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- core/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/TransportTasksActionTests.java	(revision af5fbcddfcd927838f0fc2e8ff4c25dfc23fcf92)
+++ core/src/test/java/org/elasticsearch/action/admin/cluster/node/tasks/TransportTasksActionTests.java	(revision )
@@ -755,6 +755,9 @@
             assertEquals(1, otherNode.size()); // one tasks for the all other nodes
         }
 
+        String s = response.toString();
+        logger.error("{}", s);
+
         // Group by parents
         Map<String, Object> byParent = serialize(response, new ToXContent.MapParams(Collections.singletonMap("group_by", "parents")));
         byParent = (Map<String, Object>) byParent.get("tasks");
```

and I'm now getting:

```
[2016-08-09 12:38:01,957][ERROR][org.elasticsearch.action.admin.cluster.node.tasks] Error building toString out of XContent: com.fasterxml.jackson.core.JsonGenerationException: Can not write a field name, expecting a value
	at com.fasterxml.jackson.core.JsonGenerator._reportError(JsonGenerator.java:1886)
	at com.fasterxml.jackson.core.json.UTF8JsonGenerator.writeFieldName(UTF8JsonGenerator.java:185)
	at org.elasticsearch.common.xcontent.json.JsonXContentGenerator.writeFieldName(JsonXContentGenerator.java:159)
	at org.elasticsearch.common.xcontent.XContentBuilder.field(XContentBuilder.java:198)
	at org.elasticsearch.common.xcontent.XContentBuilder.startObject(XContentBuilder.java:145)
	at org.elasticsearch.action.admin.cluster.node.tasks.list.ListTasksResponse.toXContent(ListTasksResponse.java:161)
	at org.elasticsearch.common.Strings.toString(Strings.java:901)
	at org.elasticsearch.common.Strings.toString(Strings.java:887)
	at org.elasticsearch.action.admin.cluster.node.tasks.list.ListTasksResponse.toString(ListTasksResponse.java:208)
	at org.elasticsearch.action.admin.cluster.node.tasks.TransportTasksActionTests.testTasksToXContentGrouping(TransportTasksActionTests.java:758)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1764)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:871)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:907)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:921)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:367)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:809)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:460)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:880)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:781)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:816)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:827)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:367)
	at java.lang.Thread.run(Thread.java:745)
```

@nik9000 or others: Any idea what is causing that?

**Comment by nik9000** [comment]
> Any idea what is causing that?

Sorry it took me so long to get back to you here! I *think* this is caused by `ToXContent` having a super weak contract. Some implementers spit out the *contents* of the object that they represent (anything extending from ActionResponse, QueryBuilders, etc) so they need `Strings.toString(this, true)`. Some implementers spit out a full object and they work with `String.toString(this)`.

I got the sense that lots of implementers were actually the latter, but now when I looked around I keep seeing many of the former.

Upshot: if you add a `toString` using `Strings.toString` you should add a test for the `toString` just to make sure it doesn't totally fail. I've not been good about that in the past.

**Comment by dadoonet** [comment]
@nik9000 I ran `git bisect` with a new test to find where this "regression" has been introduced:

```java
public class BisectTests extends ESTestCase {

    public void testToStringNoTask() {
        ListTasksResponse tasksResponse = new ListTasksResponse(Collections.emptyList(), Collections.emptyList(), Collections.emptyList());
        DiscoveryNodes nodes = DiscoveryNodes.builder().build();
        tasksResponse.setDiscoveryNodes(nodes);
        String string = tasksResponse.toString();
        assertThat(string, Matchers.containsString("nodes"));
    }
}
```

Apparently this regression started from 841d5a210e12bc629336396893e18872d2b1f46f (see #18939).
@tlrx might help then... :)

After this commit, test gives:

```
  2> REPRODUCE WITH: gradle :core:test -Dtests.seed=D9546C36A833B4AA -Dtests.class=org.elasticsearch.tasks.BisectTests -Dtests.method="testToStringNoTask" -Dtests.security.manager=true -Dtests.locale=sr-Latn-ME -Dtests.timezone=Pacific/Pohnpei
FAILURE 0.27s | BisectTests.testToStringNoTask <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: 
   > Expected: a string containing "nodes"
   >      but: was "Error building toString out of XContent: com.fasterxml.jackson.core.JsonGenerationException: Can not write a field name, expecting a value
   > 	at com.fasterxml.jackson.core.JsonGenerator._reportError(JsonGenerator.java:1886)
   > 	at com.fasterxml.jackson.core.json.UTF8JsonGenerator.writeFieldName(UTF8JsonGenerator.java:185)
   > 	at org.elasticsearch.common.xcontent.json.JsonXContentGenerator.writeFieldName(JsonXContentGenerator.java:159)
   > 	at org.elasticsearch.common.xcontent.XContentBuilder.field(XContentBuilder.java:198)
   > 	at org.elasticsearch.common.xcontent.XContentBuilder.startObject(XContentBuilder.java:145)
   > 	at org.elasticsearch.action.admin.cluster.node.tasks.list.ListTasksResponse.toXContent(ListTasksResponse.java:161)
   > 	at org.elasticsearch.common.Strings.toString(Strings.java:901)
   > 	at org.elasticsearch.common.Strings.toString(Strings.java:887)
   > 	at org.elasticsearch.action.admin.cluster.node.tasks.list.ListTasksResponse.toString(ListTasksResponse.java:208)
   > 	at org.elasticsearch.tasks.BisectTests.testToStringNoTask(BisectTests.java:35)
   > 	at java.lang.Thread.run(Thread.java:745)
   > "
   > 	at __randomizedtesting.SeedInfo.seed([D9546C36A833B4AA:A689B7410CCE13A8]:0)
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.elasticsearch.tasks.BisectTests.testToStringNoTask(BisectTests.java:36)
   > 	at java.lang.Thread.run(Thread.java:745)
```

**Comment by tlrx** [comment]
> Apparently this regression started from 841d5a2 (see #18939).
> @tlrx might help then... :)

This is not a regression, Jackson 2.8.1 is stricter and does not allow to write a field name if no object has been started (ie. startObject() not called before .field()).

In your case it comes from the ToXContent.toString() method which just outputs the ToXContent object, and your object needs to wrapped into another object (I suppose ListTasksResponse.toXContent() method starts directly by writing field name).

You can use the static method `Strings.toString(tasksResponse, true)` instead of `tasksResponse.toString()`in your test.

**Comment by dadoonet** [comment]
I meant it's a regression in term of our code base. `toString()` call was working before the merge of #18939 and is not working properly after that.
We just not detected it (lack of tests).

I can confirm that using:

```java

**Comment by dadoonet** [comment]
@nik9000 Thanks to @tlrx it's fixed now. Could you review the change then?

**Comment by nik9000** [comment]
> Could you review the change then?

Sure! Thanks.

File: core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java [hunk 1]
```diff
@@ -54,12 +54,18 @@
     private DiscoveryNodes discoveryNodes;
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/list/ListTasksResponse.java]: Can you make this final now?

**Comment by nik9000** [comment]
Awesome! Left a small tiny suggestion for improvement. If you can get it, great. If not, fine by me. This is much better. Thanks!

LGTM

**Comment by nik9000** [comment]
Awesome! LGTM

**Pull request merged**